lua/lvim/updater.lua

```diff
@@ -0,0 +1,298 @@
+  self.time = uv.hrtime()
+end
+function timer:stop()
+  return (uv.hrtime() - self.time) * 1e-6
+end
+timer.__index = timer
+
+local function mkdir_tmp()
+  local tmpFilePath = os.tmpname()
+  vim.fn.delete(tmpFilePath)
+
+  local sep_idx = tmpFilePath:reverse():find(path_sep)
+  local path = tmpFilePath:sub(1, #tmpFilePath - sep_idx)
+  uv.fs_mkdtemp(path .. path_sep .. "lvim_core_dl_XXXXXX")
+  return path
+end
+
+local function get_lvim_after_user_config()
+  local original_lvim = lvim
+  local user_lvim = vim.deepcopy(lvim)
+  local original_package_loaded = package.loaded
+  local user_package_loaded = {}
+  _G.lvim = user_lvim
+  _G.package.loaded = user_package_loaded
+  local ok, err = pcall(dofile, require("lvim.config"):get_user_config_path())
+  if not ok then
+    print(err)
+  end
+  _G.lvim = original_lvim
+  _G.package.loaded = original_package_loaded
+
+  return user_lvim
+end
+
+local function plug_extracted_dir(plug)
+  local commit = plug.commit
+  local repo = plug[1]
+  local name = repo:match "/(%S*)"
+  return name .. "-" .. commit
+end
+
+---downloads and installs from a core plugin entry
+---@param plug table plugin entry
+---@return function
+local function download_and_install(plug, core_install_dir, download_dir)
+  local commit = plug.commit
+  local repo = plug[1]
+  local zip_name = commit .. ".zip"
+  local extracted_dir = join_paths(core_install_dir, plug_extracted_dir(plug))
+
+  return async(function()
+    if not plug.commit then
+      error("commit missing for plugin: " .. repo)
+    end
+    local _, dir_exists = await(async_stat(extracted_dir))
+    local r = result.ok()
+    -- skip plugins that are already installed
+    if dir_exists then
+      return r
+    end
+
+    local url = "https://github.com/" .. repo .. "/archive/" .. zip_name
+
+    return r
+      :and_then(await, jobs.run({ "curl", "-LO", url }, { cwd = download_dir }))
+      :and_then(await, jobs.run({ "unzip", "-o", join_paths(download_dir, zip_name), "-d", core_install_dir }, {}))
+      :or_else(function()
+        error("download and install failed for plugin '" .. repo .. "'")
+      end)
+  end)
+end
+
+local function opendir(dir, callback)
+  uv.fs_opendir(dir, callback, 9999)
+end
+
+local async_opendir = a.wrap(opendir)
+
+-- don't want to remove any unintended directory
+local function safer_rm_rf(parent_dir, child_dir)
+  local path_to_remove = parent_dir .. path_sep .. child_dir
+  if parent_dir == nil or #parent_dir == 0 then
+    return
+  end
+  if child_dir == nil or #child_dir == 0 then
+    return
+  end
+
+  vim.fn.delete(path_to_remove, "rf")
+end
+
+local function cleanup_stale(core_install_dir, core_plugins)
+  return async(function()
+    local err, dir = await(async_opendir(core_install_dir))
+    assert(not err, "unable to open core plugin install directory")
+    local entries
+    err, entries = await(async_readdir(dir))
+    assert(not err, "unable to read core plugin install directory")
```

Comment:
Should we wrap this `assert` to include an error handler (which may decide to do `cquit`) and potentially dump the error string to some logfile, either the default one, or a temporary one?

Reply:
These are fatal errors, if they occur, they are logged to the console and the process exits.
Mind you, this is _before_ the logging facilities with structlog are available. I also don't see why logging this to a file is necessary in any way.

Reply:
I'm mostly thinking about the first part, how did you make them to be fatal? A normal `assert` will just raise an exception without really killing the process.

Reply:
It will raise an error, yes. Since there is no pcall wrapping it, and there shouldn't be, that causes it to terminate.

This is the point of an assertion, it asserts that the condition is valid and otherwise will not continue executing the code that follows.

Very simple way of testing this is to make the directory then remove permissions to read the directory.

Reply:
What I mean is that it will not terminate the actual neovim process, so the CI workflow will get stuck because neovim isn't exiting.

```lua
-- add this to your init.lua
assert(os.getenv "MY_ENV", "missing env")

vim.schedule(function()
  print "hello"
end)
```

Reply:
I assumed incorrectly that like for other print statements the headless fix would prevent the short message handler. I guess ultimately even with patching some functions, [the neovim bug remains forever punted](https://github.com/neovim/neovim/issues/9358)

I'll just make an assert_or_fail function that exits with 1 after echoing the error and use that.

Thanks for pointing this out.